subprojects/messaging/src/main/java/org/gradle/internal/event/DefaultListenerManager.java

```diff
@@ -48,46 +49,58 @@ private DefaultListenerManager(DefaultListenerManager parent) {
     }
 
     public void addListener(Object listener) {
+        ListenerDetails details = null;
         synchronized (lock) {
             if (!allListeners.containsKey(listener)) {
-                ListenerDetails details = new ListenerDetails(listener);
+                details = new ListenerDetails(listener);
                 allListeners.put(listener, details);
-                for (EventBroadcast<?> broadcaster : broadcasters.values()) {
-                    broadcaster.maybeAdd(details);
-                }
+            }
+        }
+        if (details != null) {
+            for (EventBroadcast<?> broadcaster : broadcasters.values()) {
+                broadcaster.maybeAdd(details);
             }
         }
     }
 
     public void removeListener(Object listener) {
+        ListenerDetails details;
         synchronized (lock) {
-            ListenerDetails details = allListeners.remove(listener);
+            details = allListeners.remove(listener);
             if (details != null) {
                 details.disconnect();
+            }
+        }
+        if (details != null) {
+            // block until the listener has finished notifying.
+            details.notifyingLock.lock();
+            try {
```

Comment:
⭕️ This logic would probably be better encapsulated in the ListenerDetails class, so we don't need to expose/touch its lock.